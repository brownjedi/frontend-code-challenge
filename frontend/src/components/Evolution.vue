<template>
  <div :class="styles.container">
    <transition name="fade">
      <!-- Loading -->
      <div v-if="loading" :class="styles.loading">
        <img src="/img/pokeball_bw.svg" alt="loading" />
      </div>

      <!-- Error -->
      <div v-else-if="error" :class="styles.error">An error occured</div>

      <!-- Result -->
      <div v-show="imgLoaded" v-else-if="pokemon" :class="styles.result" :key="pokemon.id" @click="navigateToPokemon">
        <img :src="pokemon.image" :alt="name" @load="onLoad" />
        <p :class="styles.id">#{{ pokemon.id }}</p>
        <p :class="styles.name">{{ pokemon.name }}</p>
      </div>
      <!-- No result -->
      <div v-else :class="styles['no-result']">No result :(</div>
    </transition>
  </div>
</template>

<script lang="ts">
import { defineComponent, ref, useCssModule, watch } from 'vue'
import { useRouter } from 'vue-router'
import { usePokemonByNameQuery } from '@/composables/pokemonByNameQuery'
export default defineComponent({
  props: {
    id: {
      type: String,
      required: true,
    },
    name: {
      type: String,
      required: true,
    },
  },
  setup(props) {
    const styles = useCssModule()
    const router = useRouter()
    const imgLoaded = ref(false)
    const { pokemon, loading, error } = usePokemonByNameQuery(props.name.toLowerCase())

    watch(loading, val => !!val && (imgLoaded.value = false))

    function onLoad() {
      imgLoaded.value = true
    }

    function navigateToPokemon() {
      router.push(`/${props.name.toLowerCase()}`)
    }
    return { styles, pokemon, loading, error, navigateToPokemon, onLoad, imgLoaded }
  },
})
</script>

<style lang="scss" module>
.container {
}

.loading,
.result {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  img {
    display: block;
    max-width: 8rem;
    height: auto;
    width: 100%;
  }

  p {
    margin: 0;
  }
}

.result {
  cursor: pointer;

  .id {
    margin-top: 1rem;
    margin-bottom: 0.35rem;
    font-weight: 500;
  }
}

.no-result {
  display: flex;
}

.error {
  color: red;
}
</style>
