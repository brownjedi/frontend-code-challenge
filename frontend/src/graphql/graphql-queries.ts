import { gql } from '@apollo/client/core'

const POKEMON_OVERVIEW_FRAGMENT = gql`
  fragment PokemonOverviewFields on Pokemon {
    id
    name
    image
    sprites {
      normal
      animated
    }
    classification
    types
    isFavorite
  }
`

export const POKEMONS_QUERY = gql`
  ${POKEMON_OVERVIEW_FRAGMENT}
  query pokemons($query: PokemonsQueryInput! = { limit: 30, offset: 0 }) {
    pokemons(query: $query) {
      limit
      offset
      count
      edges {
        ...PokemonOverviewFields
      }
    }
  }
`

export const POKEMON_TYPES_QUERY = gql`
  query pokemonTypes {
    pokemonTypes
  }
`

export const POKEMON_BY_NAME_QUERY = gql`
  ${POKEMON_OVERVIEW_FRAGMENT}
  query pokemon($name: String!) {
    pokemon: pokemonByName(name: $name) {
      ...PokemonOverviewFields
      number
      maxCP
      maxHP
      fleeRate
      resistant
      sound
      types
      weaknesses
      rgbaster @client {
        backgroundColor
        textColor
      }
      attacks {
        special {
          name
          type
          damage
        }
        fast {
          name
          type
          damage
        }
      }
      evolutionRequirements {
        name
        amount
      }
      evolutions {
        id
        name
      }
      height {
        minimum
        maximum
      }
      weight {
        minimum
        maximum
      }
    }
  }
`

export const FAVORITE_POKEMON_MUTATION = gql`
  mutation favoritePokemon($id: ID!) {
    pokemon: favoritePokemon(id: $id) {
      id
      isFavorite
    }
  }
`

export const UNFAVORITE_POKEMON_MUTATION = gql`
  mutation unFavoritePokemon($id: ID!) {
    pokemon: unFavoritePokemon(id: $id) {
      id
      isFavorite
    }
  }
`
