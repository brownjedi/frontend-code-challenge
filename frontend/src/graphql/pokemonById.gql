query pokemon($id: ID!) {
  pokemon: pokemonById(id: $id) {
    id
    name
    number
    classification
    image
    isFavorite
    maxCP
    maxHP
    fleeRate
    resistant
    sound
    types
    weaknesses
    attacks {
      special {
        name
        type
        damage
      }
      fast {
        name
        type
        damage
      }
    }
    evolutionRequirements {
      name
      amount
    }
    evolutions {
      id
      name
    }
    height {
      minimum
      maximum
    }
    weight {
      minimum
      maximum
    }
  }
}
