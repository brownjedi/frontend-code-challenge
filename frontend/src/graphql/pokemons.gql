query pokemons {
  pokemons(query: {}) {
    limit
    offset
    edges {
      id
      name
    }
  }
}
