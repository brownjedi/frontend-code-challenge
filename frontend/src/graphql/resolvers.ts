import Color from 'color'
import rgbaster from 'rgbaster'
import { PokemonInfo, RGBaster } from '@/types'

export const resolvers = {
  Pokemon: {
    rgbaster: async (_root: Pick<PokemonInfo, 'image'>): Promise<RGBaster> => {
      const rgbasterResult = await rgbaster(_root.image)
      const backgroundColor = rgbasterResult?.[0].color ?? '#FFFFFF'
      const textColor = new Color(backgroundColor).isLight() ? '#000000' : '#FFFFFF'
      return { backgroundColor, textColor }
    },
  },
}
