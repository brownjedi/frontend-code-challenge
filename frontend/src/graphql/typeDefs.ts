import { gql } from '@apollo/client/core'

export const typeDefs = gql`
  extend type Pokemon {
    rgbaster: RGBaster!
  }

  type RGBaster {
    backgroundColor: String
    textColor: String
  }
`
