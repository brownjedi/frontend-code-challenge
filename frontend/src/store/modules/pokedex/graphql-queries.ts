import gql from 'graphql-tag'

const POKEMON_OVERVIEW_FRAGMENT = gql`
  fragment PokemonOverviewFields on Pokemon {
    id
    name
    image
    sprites {
      normal
      animated
    }
    classification
    types
    isFavorite
  }
`

export const POKEMONS_QUERY = gql`
  ${POKEMON_OVERVIEW_FRAGMENT}
  query pokemons {
    pokemons(query: { limit: 100 }) {
      limit
      offset
      edges {
        ...PokemonOverviewFields
      }
    }
  }
`

export const POKEMON_BY_NAME_QUERY = gql`
  ${POKEMON_OVERVIEW_FRAGMENT}
  query pokemon($name: String!) {
    pokemon: pokemonByName(name: $name) {
      ...PokemonOverviewFields
      number
      maxCP
      maxHP
      fleeRate
      resistant
      sound
      types
      weaknesses
      attacks {
        special {
          name
          type
          damage
        }
        fast {
          name
          type
          damage
        }
      }
      evolutionRequirements {
        name
        amount
      }
      evolutions {
        id
        name
      }
      height {
        minimum
        maximum
      }
      weight {
        minimum
        maximum
      }
    }
  }
`
