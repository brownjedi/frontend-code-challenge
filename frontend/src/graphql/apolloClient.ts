/* eslint-disable graphql/template-strings */
import { PokemonList } from '@/types'
import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client/core'
import { resolvers } from './resolvers'
import { typeDefs } from './typeDefs'
import produce from 'immer'

// HTTP connection to the API
const httpLink = createHttpLink({
  // You should use an absolute URL here
  uri: 'http://localhost:4000/graphql',
})

// Cache implementation
const cache = new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        pokemons: {
          keyArgs: ['query', ['search', 'filter', ['type', 'isFavorite']]],
          merge(existing: PokemonList, incoming: PokemonList, { args }) {
            if (!existing) {
              return incoming
            }

            const offset = args?.query?.offset || 0
            const nextState = produce(existing, draftState => {
              for (let i = 0; i < incoming.edges.length; ++i) {
                draftState.edges[offset + i] = incoming.edges[i]
              }
            })

            return nextState
          },
        },
      },
    },
  },
})

// Create the apollo client
export const apolloClient = new ApolloClient({
  link: httpLink,
  cache,
  typeDefs,
  resolvers,
  defaultOptions: {
    watchQuery: {
      refetchWritePolicy: 'overwrite',
    },
  },
})
