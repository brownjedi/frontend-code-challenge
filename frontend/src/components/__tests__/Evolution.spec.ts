import { flushPromises, shallowMount } from '@vue/test-utils'
import Evolution from '@/components/Evolution.vue'
import { createMockClient, MockApolloClient } from 'mock-apollo-client'
import { POKEMON_BY_NAME_QUERY } from '@/graphql/graphql-queries'
import { provideApolloClient } from '@vue/apollo-composable'

const pokemon = {
  id: '001',
  name: 'Bulbasaur',
  image: 'http://localhost:4000/img/artwork/vector/large/bulbasaur.png',
  sprites: {
    normal: 'http://localhost:4000/img/sprites/black-white/normal/bulbasaur.png',
    animated: 'http://localhost:4000/img/sprites/black-white/anim/normal/bulbasaur.gif',
  },
  classification: 'Seed PokÃ©mon',
  types: ['Grass', 'Poison'],
  isFavorite: false,
  number: 1,
  maxCP: 951,
  maxHP: 1071,
  fleeRate: 0.1,
  resistant: ['Water', 'Electric', 'Grass', 'Fighting', 'Fairy'],
  sound: 'http://localhost:4000/sounds/1',
  weaknesses: ['Fire', 'Ice', 'Flying', 'Psychic'],
  attacks: {
    special: [
      {
        name: 'Power Whip',
        type: 'Grass',
        damage: 70,
      },
      {
        name: 'Seed Bomb',
        type: 'Grass',
        damage: 40,
      },
      {
        name: 'Sludge Bomb',
        type: 'Poison',
        damage: 55,
      },
    ],
    fast: [
      {
        name: 'Tackle',
        type: 'Normal',
        damage: 12,
      },
      {
        name: 'Vine Whip',
        type: 'Grass',
        damage: 7,
      },
    ],
  },
  evolutionRequirements: {
    name: 'Bulbasaur candies',
    amount: 25,
  },
  evolutions: [
    {
      id: '002',
      name: 'Ivysaur',
    },
    {
      id: '003',
      name: 'Venusaur',
    },
  ],
  height: {
    minimum: '0.61m',
    maximum: '0.79m',
  },
  weight: {
    minimum: '6.04kg',
    maximum: '7.76kg',
  },
}

const routerPushMock = jest.fn()

jest.mock('vue-router', () => ({
  useRouter: () => ({
    push: routerPushMock,
  }),
}))

describe('components/Evolution.vue', () => {
  let requestHandler: jest.Mock
  let mockClient: MockApolloClient

  beforeEach(() => {
    jest.resetAllMocks()
  })

  beforeEach(() => {
    mockClient = createMockClient()
    requestHandler = jest.fn()
    mockClient.setRequestHandler(POKEMON_BY_NAME_QUERY, requestHandler)
    provideApolloClient(mockClient)
  })

  describe('when query loading', () => {
    it('has the expected html structure', () => {
      const wrapper = shallowMount(Evolution, {
        props: {
          id: pokemon.id,
          name: pokemon.name,
        },
      })
      expect(wrapper.element).toMatchSnapshot()
    })

    it('sends the pokemon name in the graphql query', () => {
      shallowMount(Evolution, {
        props: {
          id: pokemon.id,
          name: pokemon.name,
        },
      })

      expect(requestHandler).toHaveBeenCalledWith({ name: pokemon.name.toLowerCase() })
    })
  })

  describe('when query fails', () => {
    it('shows error message', async () => {
      requestHandler.mockRejectedValueOnce(new Error('Graphql query failed'))

      const wrapper = shallowMount(Evolution, {
        props: {
          id: pokemon.id,
          name: pokemon.name,
        },
      })

      expect(requestHandler).toHaveBeenCalledTimes(1)

      // wait for apollo results
      await flushPromises()

      // ensure that result container exists
      expect(wrapper.find('.error').exists()).toBe(true)
      expect(wrapper.find('.error').text()).toBe('An error occured')
    })
  })

  describe('when query succeeds', () => {
    it('shows no result message when no data', async () => {
      requestHandler.mockResolvedValueOnce({
        data: { pokemon: null },
      })

      const wrapper = shallowMount(Evolution, {
        props: {
          id: pokemon.id,
          name: pokemon.name,
        },
      })

      expect(requestHandler).toHaveBeenCalledTimes(1)

      // wait for apollo results
      await flushPromises()

      // ensure that result container exists
      expect(wrapper.find('.no-result').exists()).toBe(true)
      expect(wrapper.find('.no-result').text()).toBe('No result :(')
    })

    it('renders results correctly when data is present', async () => {
      requestHandler.mockResolvedValueOnce({
        data: { pokemon },
      })

      const wrapper = shallowMount(Evolution, {
        props: {
          id: pokemon.id,
          name: pokemon.name,
        },
      })

      // wait for apollo results
      await flushPromises()

      expect(wrapper.element).toMatchSnapshot()
    })

    it('navigates to pokemon info on click', async () => {
      requestHandler.mockResolvedValueOnce({
        data: { pokemon },
      })

      const wrapper = shallowMount(Evolution, {
        props: {
          id: pokemon.id,
          name: pokemon.name,
        },
      })

      // wait for apollo results
      await flushPromises()

      await wrapper.find('.result').trigger('click')

      expect(routerPushMock).toHaveBeenCalledTimes(1)
      expect(routerPushMock).toHaveBeenCalledWith(`/${pokemon.name.toLowerCase()}`)
    })
  })
})
