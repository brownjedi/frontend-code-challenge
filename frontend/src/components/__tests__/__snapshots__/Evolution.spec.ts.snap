// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Evolution.vue when query loading has the expected html structure 1`] = `
<div
  class=""
>
  <transition-stub>
    <div
      class="loading"
    >
      <img
        alt="loading"
        src="/img/pokeball_bw.svg"
      />
    </div>
  </transition-stub>
</div>
`;

exports[`components/Evolution.vue when query succeeds renders results correctly when data is present 1`] = `
<div
  class=""
>
  <transition-stub>
    <div
      class="result"
      style="display: none;"
    >
      <img
        alt="Bulbasaur"
        src="http://localhost:4000/img/artwork/vector/large/bulbasaur.png"
      />
      <p
        class="id"
      >
        #001
      </p>
      <p
        class=""
      >
        Bulbasaur
      </p>
    </div>
  </transition-stub>
</div>
`;
